python/private/attr_builders.bzl

```diff
@@ -0,0 +1,1284 @@
+# Copyright 2025 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Builders for creating attributes et al."""
+
+load("@bazel_skylib//lib:types.bzl", "types")
+load(
+    ":builders_util.bzl",
+    "kwargs_getter",
+    "kwargs_set_default_doc",
+    "kwargs_set_default_ignore_none",
+    "kwargs_set_default_list",
+    "kwargs_set_default_mandatory",
+    "kwargs_setter",
+    "to_label_maybe",
+)
+
+def _kwargs_set_default_allow_empty(kwargs):
+    existing = kwargs.get("allow_empty")
+    if existing == None:
+        kwargs["allow_empty"] = True
+
+def _kwargs_set_default_allow_files(kwargs):
+    existing = kwargs.get("allow_files")
+    if existing == None:
+        kwargs["allow_files"] = False
+
+def _common_label_build(self, attr_factory):
+    kwargs = dict(self.kwargs)
+    kwargs["cfg"] = self.cfg.build()
+    return attr_factory(**kwargs)
+
+def _AttrCfg_typedef():
+    """Builder for `cfg` arg of label attributes.
+
+    :::{function} inputs() -> list[Label]
+    :::
+
+    :::{function} outputs() -> list[Label]
+    :::
+
+    :::{function} which_cfg() -> str
+
+    Tells which of the cfg modes is set. Will be one of: target, exec, none,
+    or implementation
+    :::
+    """
+
+def _AttrCfg_new(
+        inputs = None,
+        outputs = None,
+        **kwargs):
+    """Creates a builder for the `attr.cfg` attribute.
+
+    Args:
+        inputs: {type}`list[Label] | None` inputs to use for a transition
+        outputs: {type}`list[Label] | None` outputs to use for a transition
+        **kwargs: {type}`dict` Three different keyword args are supported.
+            The presence of a keyword arg will mark the respective mode
+            returned by `which_cfg`.
+            - `cfg`: string of either "target" or "exec"
+            - `exec_group`: string of an exec group name to use. None means
+              to use regular exec config (i.e. `config.exec()`)
+            - `implementation`: callable for a custom transition function.
+
+    Returns:
+        {type}`AttrCfg`
+    """
+    state = {
+        "inputs": inputs,
+        "outputs": outputs,
+        # Value depends on "which" key
+        # For which=impl, the value is a function or arbitrary object
+        "value": True,
+        # str: target, exec, none, or impl
+        "which": "target",
```

Comment:
nit: we could have an enum here? That would make the code read a little bit easier.

Reply:
Done